shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs

```diff
@@ -220,19 +220,31 @@ balance (UTxO utxo) = foldr addCoins 0 utxo
   where
     addCoins (TxOut _ a) b = a + b
 
--- | Determine the total deposit amount needed
+-- | Determine the total deposit amount needed.
+-- We rely on the dvalue to get the deposit value for each certificate,
+-- but for a list of certificates we must only count a (RegPool hk)
+-- deposit once for a particular hk.
 totalDeposits ::
   PParams ->
   StakePools crypto ->
   [DCert crypto] ->
   Coin
-totalDeposits pc (StakePools stpools) cs = foldl' f (Coin 0) cs'
+totalDeposits pc (StakePools stpools) cs = 
```

Comment:
how about something like this, for better readability:

```
totalDeposits pp (StakePools stpools) cs =
  (_keyDeposit pp) * numKeys + (_poolDeposit pp) * numNewPools
  where             
    numKeys = intToCoin . length $ filter isRegKey cs
    pools = Set.fromList . Maybe.catMaybes $ fmap getKeyHashFromRegPool cs
    numNewPools = intToCoin . length $ (Map.keysSet stpools) `Set.difference` pools
    intToCoin = Coin . toInteger
```
(`isRegKey` is in `Shelley.Spec.Ledger.Delegation.Certificates`.)

and

```
getKeyHashFromRegPool :: DCert crypto -> Maybe (KeyHash 'StakePool crypto)
getKeyHashFromRegPool (DCertPool (RegPool p)) = Just . _poolPubKey $ p
getKeyHashFromRegPool _ = Nothing
```

We could then also simplify to this `ax+by` style in the formal spec.

Reply:
That's much nicer thanks 👍